lib/Analysis/Utility.cpp

```diff
@@ -69,18 +69,22 @@ unsigned ReduceOpHelper::getThreadOffsetOnReductionAxis() {
   }
 
   unsigned threadOffset = 1;
-  if (auto sliceLayout = mlir::dyn_cast<SliceEncodingAttr>(srcLayout)) {
-    auto parentLayout = sliceLayout.getParent();
-    auto threadsPerWarp = getThreadsPerWarp(parentLayout);
-    threadOffset = threadsPerWarp[sliceLayout.getDim()];
-  } else {
-    auto threadsPerWarp = getThreadsPerWarp(srcLayout);
-    auto order = getThreadOrder(srcLayout);
-    for (unsigned i = 0; i < order.size(); i++) {
-      if (order[i] == axis)
-        break;
-      threadOffset *= threadsPerWarp[order[i]];
-    }
+  SmallVector<int> dimsRemoved;
+  while (auto sliceLayout = mlir::dyn_cast<SliceEncodingAttr>(srcLayout)) {
+    dimsRemoved.push_back(sliceLayout.getDim());
+    srcLayout = sliceLayout.getParent();
+  }
+  int adjustedAxis = axis;
+  for (auto dim : dimsRemoved) {
```

Comment:
I'm confused about `adjustedAxis`. Maybe I'm missing something.

I thought `sliceLayout.getDim()` is a relative index to its parent, so we will just need the outmost `sliceLayout` and get its dim to be `adjustedAxis`?

Reply:
this is for the destination, here we need to calculate the number of threads between a value and the next element along one dimension. For that we need to get to the most inner parent and multiply the number of threads that any dimension before the reduction dimension has. The adjustedAxis computes the axis relative to the most inner parent (axis itself is relative to the sliced tensor)
Does that make sense? I can try to summarize that into a comment if you think it is understandable

Reply:
Thanks. I get it now. It's recovering the slicing dim of the inner most slice layout with regard to the original layout